File: examples/scripts/python-orca/simulation.py
import json
import sys
from threading import Thread

import numpy as np
from mathutils import normalize
from obstacle import Obstacle
from participant import Participant
from util import log


class Simulation(Thread):
    """
    Class for communicating with the simulation. It handles serialization and deserialization of the data.
    """

    running = False

    def __init__(self):
        """
        Constructor

        Paremeters:
            fps: The frames per second of the simulation. This does not refer to the simulation ticks, but rather the "lookup range" for each participant.
        """
        super().__init__()
        self.__setup()

    def __setup(self):
        # gather setup information
        setup = json.loads(sys.stdin.readline())
        position = np.array([setup["position"]["x"], setup["position"]["y"]])
        radius = setup["radius"]
        safezone = setup["safezone"]
        target = np.array([setup["target"]["x"], setup["target"]["y"]])
        vmax = setup["vmax"]
        # create participant for this unit
        self.we = Participant(position, normalize(
            target - position) * vmax, radius, safezone, vmax, target)

    def start(self, cb):
        """
        Starts the simulation and registers the callback function, which is called every tick.
        """
        if self.running:
            # don't be rude and start the simulation twice
            raise RuntimeError("Simulation already running")
        self.cb = cb
        self.running = True
        super().start()

    def run(self):
        self.__loop(self.cb)

    def __loop(self, callback):
        """
        The main loop of the simulation
        """
        while self.running:
            # get current tick information
            inp = json.loads(sys.stdin.readline())
            position = np.array([inp["position"]["x"], inp["position"]["y"]])
            self.we.update_position(position)

            # read information about all other participants
            participants = []
            for p in inp["participants"]:
                participant = Participant(np.array([p["position"]["x"], p["position"]["y"]]), np.array(
                    [p["velocity"]["x"], p["velocity"]["y"]]), p["radius"], p["safezone"])
                participants.append(participant)
            obstacles = []
            for o in inp["obstacles"]:
                obstacles.append(Obstacle(o))

            # call the callback function
            velocity = callback(self.we, participants, obstacles)
            # send the new velocity to the simulation
            val = {
                "action": "move",
                "payload": {
                    "x": velocity[0],
                    "y": velocity[1],
                },
                "debug": inp["participants"]
            }
            print(json.dumps(val))

    def log(self, msg):
        """
        Log a message to the simulation.
        """
        log("log", msg)

    def stop(self):
        """
        Stop the simulation.
        """
        # set flag and send stop message
        self.running = False
        val = {
            "action": "stop",
            "payload": {}
        }
        print(json.dumps(val))

File: examples/scripts/rust-orca/src/simulation.rs
use std::io;

use orca_rs::{ndarray::arr1, obstacle::Obstacle, participant::Participant};
use serde_json::{json, Value};

pub struct Simulation {
    we: Participant,
    tau: f64,
}

#[cfg(feature = "confidence")]
const CONF: f64 = 2.0;

#[cfg(not(feature = "confidence"))]
const CONF: f64 = 0.0;

impl Simulation {
    pub fn new() -> Simulation {
        let mut buffer = String::new();
        io::stdin()
            .read_line(&mut buffer)
            .expect("Error reading stdio");

        // Read setup information from stdin
        let setup: Value =
            serde_json::from_str(buffer.as_str()).expect("Error decoding setup message");
        let position = arr1(&[
            setup["position"]["x"].as_f64().unwrap(),
            setup["position"]["y"].as_f64().unwrap(),
        ]);
        let target = arr1(&[
            setup["target"]["x"].as_f64().unwrap(),
            setup["target"]["y"].as_f64().unwrap(),
        ]);
        let vmax = setup["vmax"].as_f64().unwrap();
        let tau = setup["tau"].as_f64().unwrap();
        let radius = setup["radius"].as_f64().unwrap();
        let we = Participant::new(position, arr1(&[0.0, 0.0]), radius, 0.0)
            .with_inner_state(vmax, target);

        Simulation { we, tau }
    }

    /// Start this simulation with a callback to call during each tick
    pub fn start(&mut self) {
        // continuously perform ticks
        loop {
            let mut buffer = String::new();
            // read information from stdin
            io::stdin()
                .read_line(&mut buffer)
                .expect("Error reading stdin");
            let inp: Value = serde_json::from_str(buffer.trim())
                .expect("Error decoding message from simulation.");

            // update our information
            let position = arr1(&[
                inp["position"]["x"].as_f64().unwrap(),
                inp["position"]["y"].as_f64().unwrap(),
            ]);
            self.we.update_position(&position);
            self.we.confidence = inp["stddev"].as_f64().unwrap() * CONF + 0.001;

            // convert the information about all other participants
            let mut participants: Vec<Participant> = Vec::new();
            for p in inp["participants"].as_array().unwrap() {
                let position = arr1(&[
                    p["position"]["x"].as_f64().unwrap(),
                    p["position"]["y"].as_f64().unwrap(),
                ]);
                let velocity = arr1(&[
                    p["velocity"]["x"].as_f64().unwrap(),
                    p["velocity"]["y"].as_f64().unwrap(),
                ]);
                let radius = p["radius"].as_f64().unwrap();
                let confidence = p["stddev"].as_f64().unwrap() * CONF + 0.001;
                participants.push(Participant::new(position, velocity, radius, confidence))
            }

            // get information about static obstacles
            let mut obstacles: Vec<Obstacle> = Vec::new();
            for o in inp["obstacles"].as_array().unwrap() {
                obstacles.push(Obstacle {
                    start: arr1(&[
                        o["start"]["x"].as_f64().unwrap(),
                        o["start"]["y"].as_f64().unwrap(),
                    ]),
                    end: arr1(&[
                        o["end"]["x"].as_f64().unwrap(),
                        o["end"]["y"].as_f64().unwrap(),
                    ]),
                    radius: 0.0,
                });
            }
            // get new velocity from callback & send it via stdout
            let vel = self.we.orca(&mut participants, &obstacles, self.tau);
            let val = json!({
                "action": "move",
                "payload": {
                    "x": vel[0],
                    "y": vel[1]
                }
            });
            println!("{}", val);
        }
    }
}
